### PR TITLE
add proposed pypy versions to matrix

### PR DESCRIPTION
**Pull request by msarahan**
ping @ocefpaf @mattip

### CODE REVIEW

**Comment by scopatz** [comment]
LGTM

**Comment by isuruf** [comment]
How does this work? Are there conda packages named `python` with version `pypy.2.7.7.1`?

Should this be done using a migrator instead of enabling this for all feedstocks at once?

**Comment by ocefpaf** [comment]
> Should this be done using a migrator instead of enabling this for all feedstocks at once?

I believe that the next step, before we merge this one, is a migrator. However, I think that the plan is to try 1-3 feedstocks first with this PR.

I'll be sprinting on ESIP stuff today but I'll try to give this a go.

**Comment by mattip** [comment]
@ocefpaf is there something I can do to "try 1-3 feedstocks first with this PR"?

**Comment by ocefpaf** [comment]
> @ocefpaf is there something I can do to "try 1-3 feedstocks first with this PR"?

@mattip you can try to install conda-smithy from this PR and re-render a feedstock. There are probably more details we need to solve though. I'll try to tackle this over the weekend. (I'm still trying to catch up with my e-mails after PyCon :weary:)

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -543,6 +543,9 @@ python:
   - 2.7
   - 3.6
   - 3.7
+  # pypy only supports win32 so far
+  - pypy.3.6.7.1  # [unix or win32]
+  - pypy.2.7.7.1  # [unix or win32]
```
Comment (ocefpaf) [inline recipe/conda_build_config.yaml]: We need to build `pypy` with those version strings first, right?

Comment (mattip) [reply recipe/conda_build_config.yaml]: Sorry, I am a conda newbie. What exactly do I need to do to 
- "install conda-smithy from this PR and re-render a feedstock"

Is there a "getting started" page somewhere?

Comment (ocefpaf) [reply recipe/conda_build_config.yaml]: Sorry I was not clear. You would need a development install of the package with this PR. I usually clone the repo, checkout the branch, and do `pip install -e . --no-deps --force-reinstall`. However, there are a lot of details I left out in order to build a feedstock locally. Some of those are listed in our docs: https://conda-forge.org/docs/maintainer/updating_pkgs.html?#testing-changes-locally

I guess that the best course of action is for us, `conda-forge`, to vote on this PR and merge it ASAP. I'll try to run a few tests myself this weekend and I'll get back to you.

Comment (mattip) [reply recipe/conda_build_config.yaml]: @ocefpaf did you have a chance to take a look?

Comment (ocefpaf) [reply recipe/conda_build_config.yaml]: Sorry, no. Kind of swamped with the day job and some travels in the agenda. But I did not forget this and it is part of `conda-forge`'s plan to roll this out ASAP.

Comment (mattip) [reply recipe/conda_build_config.yaml]: ping

Comment (CJ-Wright) [inline recipe/conda_build_config.yaml]: Do we have win32 builds?

Comment (mattip) [reply recipe/conda_build_config.yaml]: PyPy has only win32, no win64 at this time

**Comment by CJ-Wright** [comment]
@msarahan do you mind if I write the cfep9 version of this?

File: recipe/migrations/pypy.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    version
+  exclude:
+    - c_compiler
+    - vc
+    - cxx_compiler
+  migration_number:  # Only use this if the bot messes up, putting this in will cause a complete rerun of the migration
+    1 
+  bump_number: 0
+
+python:
```
Comment (CJ-Wright) [inline recipe/migrations/pypy.yaml]: Does this need a modification to the zip keys?

Comment (isuruf) [reply recipe/migrations/pypy.yaml]: Nope. Not building for windows here.

Comment (CJ-Wright) [reply recipe/migrations/pypy.yaml]: Ahh ok

**Comment by CJ-Wright** [comment]
How difficult of a rebuild do we expect this to be? (how often will we have to retool the recipes?) I wonder if we should try to roll out more automerge things before we set this off. (not that automerge should stop this, but it would make the maintainers' burden lower)

**Comment by isuruf** [comment]
@conda-forge/core, do we start building everything or just a few packages first and then decide later?

**Comment by jakirkham** [comment]
+1 for starting with a subset. It's unclear how much of the stack is needed yet. Besides we can always change our minds if it goes well.

**Comment by beckermr** [comment]
We don’t have scoped migrations here yet so if we want a subset we’ll need to make a pr on the bot.

**Comment by isuruf** [comment]
@beckermr, can we start slow then? 1 PR per hour with packages with more children more priority?

**Comment by jakirkham** [comment]
How did this get solved for `aarch64` and `ppc64le`?

**Comment by isuruf** [comment]
What do you mean?

**Comment by beckermr** [comment]
@jakirkham The arm and ppce migrations have a special migrator outside of the CFEP-9 migrations.

**Comment by beckermr** [comment]
@isuruf We'll need to turn down the PRs here: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/auto_tick.py#L501

@jakirkham See the special migrator here: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/migrators/arch.py#L18

We had opened an issue to generalize the aarch/ppce migrator to handle "frontier" migrations. That issue is here: https://github.com/regro/cf-scripts/issues/782

We had pypy in mind when discussing, but this issue has not been addressed yet.

**Comment by jakirkham** [comment]
Thanks for the context! I like the term frontier migrator 😄

Sounds like Isuru's solution is easiest.

**Comment by beckermr** [comment]
I think a better way to do this going forward is to support migrator options directly in the CFEP-9 YAML. We can allow people to set the number of PRs per hour, packages to ignore, a good list of feedstocks, etc.

**Comment by jakirkham** [comment]
Yeah that seems sensible to me.

IDK if you are around @CJ-Wright. If so, any thoughts on this idea?

File: recipe/migrations/pypy.yaml [hunk 3]
```diff
@@ -4,7 +4,7 @@ __migrator:
     version
   migration_number:  # Only use this if the bot messes up, putting this in will cause a complete rerun of the migration
     1
-  bump_number: 0
+  bump_number: 1   # Hashes changed for cpython, so it's better to bump build numbers.
```
Comment (beckermr) [inline recipe/migrations/pypy.yaml]: You'll need to add the PR limit in here I think.

**Comment by beckermr** [comment]
We apparently already support things like this which shows you how much I have to learn about the migrations code. ;p

**Comment by isuruf** [comment]
This is blocked by https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/70

**Comment by ocefpaf** [comment]
Looks like we are finally ready for this. Thanks @msarahan for the PR and the initial work, @mattip for the guidance and patience, and @isuruf for identifying and fixing all the issues that paved the way to merge this!

**Comment by mattip** [comment]
Is there something to document both from the perspective of a recipe writer and from the perspective of an end user? It would be nice to make this option discoverable

**Comment by beckermr** [comment]
Yes we should document in the main CF docs.

**Comment by isuruf** [comment]
See https://conda-forge.org/docs/maintainer/knowledge_base.html#pypy-builds

**Pull request merged**